crypto/ml_dsa/ml_dsa_key.c

```diff
@@ -0,0 +1,380 @@
+    ctx = ossl_ml_dsa_ctx_new(key->params->alg, key->libctx, key->propq);
+    if (ctx == NULL)
+        return 0;
+
+    vector_init(&t1, key->params->k);
+    vector_init(&t0, key->params->k);
+    if (!public_from_private(ctx, key, &t1, &t0))
+        goto err;
+
+    ret = vector_equal(&t1, &key->t1) && vector_equal(&t0, &key->t0);
+err:
+    ossl_ml_dsa_ctx_free(ctx);
+    return ret;
+}
+
+static int shake_xof(EVP_MD_CTX *ctx, const uint8_t *in, size_t in_len,
+                     uint8_t *out, size_t out_len)
+{
+    return (EVP_DigestInit_ex2(ctx, NULL, NULL) == 1
+            && EVP_DigestUpdate(ctx, in, in_len) == 1
+            && EVP_DigestFinalXOF(ctx, out, out_len) == 1);
+}
+
+/*
+ * @brief Generate a public-private key pair from a seed.
+ * See FIPS 204, Algorithm 6 ML-DSA.KeyGen_internal().
+ *
+ * @param entropy The input seed
+ * @param entropy_len The size of entropy (Should be 32 bytes)
+ *
+ *
+ * @returns 1 on success or 0 on failure.
+ */
+static int keygen_internal(ML_DSA_CTX *ctx, const uint8_t *seed, size_t seed_len,
+                           ML_DSA_KEY *out)
+{
+    int ret = 0;
+    uint8_t augmented_seed[ML_DSA_SEED_BYTES + 2];
+    uint8_t expanded_seed[ML_DSA_RHO_BYTES + ML_DSA_PRIV_SEED_BYTES + ML_DSA_K_BYTES];
+    const uint8_t *const rho = expanded_seed; /* p = Public Random Seed */
+    const uint8_t *const priv_seed = expanded_seed + ML_DSA_RHO_BYTES;
+    const uint8_t *const K = priv_seed + ML_DSA_PRIV_SEED_BYTES;
+    const ML_DSA_PARAMS *params = ctx->params;
+
+    /* augmented_seed = seed || k || l */
+    memcpy(augmented_seed, seed, seed_len);
+    augmented_seed[ML_DSA_SEED_BYTES] = (uint8_t)params->k;
+    augmented_seed[ML_DSA_SEED_BYTES + 1] = (uint8_t)params->l;
+    /* Expand the seed into p[32], p'[64], K[32] */
+    if (!shake_xof(ctx->h_ctx, augmented_seed, sizeof(augmented_seed),
+                   expanded_seed, sizeof(expanded_seed)))
+        goto err;
+
+    memcpy(out->rho, rho, sizeof(out->rho));
+    memcpy(out->K, K, sizeof(out->K));
+
+    ret = ossl_ml_dsa_sample_expandS(ctx->h_ctx, params->eta, priv_seed,
+                                     &out->s1, &out->s2)
+        && public_from_private(ctx, out, &out->t1, &out->t0)
+        && ossl_ml_dsa_pk_encode(out)
+        && shake_xof(ctx->h_ctx, out->pub_encoding, out->params->pk_len,
+                     out->tr, sizeof(out->tr))
+        && ossl_ml_dsa_sk_encode(out);
+err:
+    OPENSSL_cleanse(augmented_seed, sizeof(augmented_seed));
+    OPENSSL_cleanse(expanded_seed, sizeof(expanded_seed));
+    return ret;
+}
+
+int ossl_ml_dsa_generate_key(ML_DSA_CTX *ctx, OSSL_LIB_CTX *lib_ctx,
+                             const uint8_t *entropy, size_t entropy_len,
+                             ML_DSA_KEY *out)
+{
+    int ret = 0;
+    uint8_t seed[32];
+    size_t seed_len = sizeof(seed);
+
+    if (ctx->params != out->params)
+        return 0;
+
+    if (entropy != NULL && entropy_len != 0) {
+        if (entropy_len < seed_len)
```

Comment:
Is it correct for the caller to offer a larger seed (more entropy) than is required by the algorithm?
Perhaps providing the wrong seed size should be an error?